From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Sun, 30 Oct 2022 03:44:59 +0100
Subject: [PATCH] 2000 players entities optimization.


diff --git a/src/main/java/net/minecraft/server/level/ServerChunkCache.java b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
index 30d077fb5062970eb0985bce23faadc3693b04c9..3a0fa72f322aa60405542026b05fe75a36e5b729 100644
--- a/src/main/java/net/minecraft/server/level/ServerChunkCache.java
+++ b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
@@ -785,6 +785,7 @@ public class ServerChunkCache extends ChunkSource {
             // Paper start - optimize isOutisdeRange
             ChunkMap playerChunkMap = this.chunkMap;
             for (ServerPlayer player : this.level.players) {
+                if (MultiPaper.isExternalPlayer(player)) continue; // MultiPaper
                 if (!player.affectsSpawning || player.isSpectator()) {
                     playerChunkMap.playerMobSpawnMap.remove(player);
                     continue;
@@ -964,7 +965,7 @@ public class ServerChunkCache extends ChunkSource {
             List<net.minecraft.network.Connection> disabledFlushes = new java.util.ArrayList<>(this.level.players.size());
             for (ServerPlayer player : this.level.players) {
                 net.minecraft.server.network.ServerGamePacketListenerImpl connection = player.connection;
-                if (connection != null) {
+                if (connection != null && !MultiPaper.isExternalPlayer(player)) { // MultiPaper
                     connection.connection.disableAutomaticFlush();
                     disabledFlushes.add(connection.connection);
                 }
diff --git a/src/main/java/org/spigotmc/ActivationRange.java b/src/main/java/org/spigotmc/ActivationRange.java
index 8707382d77d627b07b4450e7d419f4c8891c8077..b6601ee3612bc4c793168460037b450a5cf59d82 100644
--- a/src/main/java/org/spigotmc/ActivationRange.java
+++ b/src/main/java/org/spigotmc/ActivationRange.java
@@ -42,6 +42,7 @@ import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import java.util.List;
 // Pufferfish end
+import puregero.multipaper.MultiPaper;
 
 public class ActivationRange
 {
@@ -197,6 +198,8 @@ public class ActivationRange
 
         for ( Player player : world.players() )
         {
+            if (MultiPaper.isExternalPlayer(player) && MinecraftServer.currentTick - player.activatedTick < 10) continue; // MultiPaper optimization
+
             player.activatedTick = MinecraftServer.currentTick;
             if ( world.spigotConfig.ignoreSpectatorActivation && player.isSpectator() )
             {
