From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Fri, 9 Dec 2022 06:06:54 +0100
Subject: [PATCH] Make ender pearls not despawn on death.


diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
index 23f77dfc6df93ef6a70e14a7e410263d39fae2cb..08957a35b87ffaf0a2d68d0ed2a0e1ddbc53c1ed 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -103,7 +103,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
         Entity entity = this.getOwner();
 
         if (entity instanceof Player && !entity.isAlive()) {
-            this.discard();
+            // this.discard(); // Xymb
         } else {
             super.tick();
         }
