From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Fri, 23 Dec 2022 06:22:52 +0100
Subject: [PATCH] Allow vertical elytra.


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index d45ed480f870550df020e36c27527ad1ccc5c2ce..dd4bf3763e794b2756bfba29f993898cd4f64174 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1676,7 +1676,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
                                     this.player.getLevel().getChunkSource().move(this.player);
                                 }
                                 this.player.doCheckFallDamage(this.player.getY() - d6, packet.isOnGround());
-                                this.player.setOnGround(packet.isOnGround());
+                                this.player.setOnGround(false);
                                 if (flag) {
                                     this.player.resetFallDistance();
                                 }
