From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Thu, 2 Feb 2023 04:44:38 +0100
Subject: [PATCH] Print the stack trace for DisconnectedReason.


diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index b4e7aba7e8e69360ead02cc836379419bf9aa5c8..625f8a041a55be56ad05853b13aafec91ea0bd5a 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -632,6 +632,10 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
         if (this.channel.isOpen()) {
             this.channel.close(); // We can't wait as this may be called from an event loop.
             this.disconnectedReason = disconnectReason;
+            if (this.disconnectedReason.getString().contains("Exception")) {
+                Connection.LOGGER.info("Disconnected reason: {}", disconnectedReason.getString()); // Xymb
+                Thread.dumpStack(); // Xymb
+            }
         }
 
     }
