From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Tue, 7 Feb 2023 16:52:19 +0100
Subject: [PATCH] Don't sendAsync commands to external players.

Optimization for TAB.

diff --git a/src/main/java/net/minecraft/commands/Commands.java b/src/main/java/net/minecraft/commands/Commands.java
index 19cb13213095a7f23ba196013214652c5824f845..f4779deb7e2e9248b87128a7f01e288d78dec029 100644
--- a/src/main/java/net/minecraft/commands/Commands.java
+++ b/src/main/java/net/minecraft/commands/Commands.java
@@ -393,6 +393,7 @@ public class Commands {
     );
 
     private void sendAsync(ServerPlayer player) {
+        if (puregero.multipaper.MultiPaper.isExternalPlayer(player)) return; // MultiPaper
         // Paper end - Async command map building
         Map<CommandNode<CommandSourceStack>, CommandNode<SharedSuggestionProvider>> map = Maps.newIdentityHashMap(); // Use identity to prevent aliasing issues
         RootCommandNode vanillaRoot = new RootCommandNode();
