From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Mon, 29 Aug 2022 23:48:18 +0200
Subject: [PATCH] Add "don'tt spill contents of shulker boxes" optimization
 option.

Signed-off-by: Xymb <xymb@endcrystal.me>

diff --git a/src/main/java/net/minecraft/world/item/BlockItem.java b/src/main/java/net/minecraft/world/item/BlockItem.java
index 62d6c5b7590ff4faef5d8c7a8be03155b7338480..ca4ebc35f25bd86310b6d8f689889d442e245b42 100644
--- a/src/main/java/net/minecraft/world/item/BlockItem.java
+++ b/src/main/java/net/minecraft/world/item/BlockItem.java
@@ -35,6 +35,7 @@ import net.minecraft.world.phys.shapes.CollisionContext;
 import org.bukkit.craftbukkit.block.CraftBlock;
 import org.bukkit.craftbukkit.block.data.CraftBlockData;
 import org.bukkit.event.block.BlockCanBuildEvent;
+import puregero.multipaper.config.MultiPaperConfiguration;
 // CraftBukkit end
 
 public class BlockItem extends Item {
@@ -294,7 +295,7 @@ public class BlockItem extends Item {
 
     @Override
     public void onDestroyed(ItemEntity entity) {
-        if (this.block instanceof ShulkerBoxBlock) {
+        if (this.block instanceof ShulkerBoxBlock && MultiPaperConfiguration.get().optimizations.shulkerBoxItemDropContentsWhenDestroyed) {
             ItemStack itemstack = entity.getItem();
             CompoundTag nbttagcompound = BlockItem.getBlockEntityData(itemstack);
 
diff --git a/src/main/java/puregero/multipaper/config/MultiPaperConfiguration.java b/src/main/java/puregero/multipaper/config/MultiPaperConfiguration.java
index de7d859b9de4d34e834cae108c4f796e92ee67c0..85a0669d627fc2cecb6eacec954a79ba1ceb1e3f 100644
--- a/src/main/java/puregero/multipaper/config/MultiPaperConfiguration.java
+++ b/src/main/java/puregero/multipaper/config/MultiPaperConfiguration.java
@@ -61,6 +61,7 @@ public class MultiPaperConfiguration extends ConfigurationPart {
         public boolean reducePlayerPositionUpdatesInUnloadedChunks = false;
         public int ticksPerInactiveEntityTracking = 1;
         public boolean useEventBasedIo = true;
+        public boolean shulkerBoxItemDropContentsWhenDestroyed = true;
     }
 
     public SyncSettings syncSettings;
