From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Mon, 24 Oct 2022 01:51:43 +0200
Subject: [PATCH] Stop "Channel is closed" spam


diff --git a/src/main/java/puregero/multipaper/ExternalServerConnection.java b/src/main/java/puregero/multipaper/ExternalServerConnection.java
index 9b9376f0b775d39a18f69200a0ef839a9f46c7c7..4107d90a1283619ed75aae0c6c3562fffe4072bb 100644
--- a/src/main/java/puregero/multipaper/ExternalServerConnection.java
+++ b/src/main/java/puregero/multipaper/ExternalServerConnection.java
@@ -177,7 +177,7 @@ public class ExternalServerConnection extends ChannelInitializer<SocketChannel>
         long start = System.nanoTime();
 
         if (!channel.isOpen()) {
-            new IOException("Channel is closed for " + externalServer.getName()).printStackTrace();
+//            new IOException("Channel is closed for " + externalServer.getName()).printStackTrace();
         } else {
             onConnect.thenRun(() -> {
                 if (channel.eventLoop().inEventLoop()) {
