From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Mon, 24 Oct 2022 01:10:28 +0200
Subject: [PATCH] Server now stops correctly after crashing.


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 5d5d15e2723643694c5f36d43c50c01eaa5952c4..fb1ae2f92461d5671c19521e0fd795d8c14487b7 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -949,7 +949,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         this.isSaving = true;
         if (this.playerList != null) {
             MinecraftServer.LOGGER.info("Saving players");
-            this.playerList.saveAll();
+            this.playerList.saveAll(-1, false); // MultiPaper
             this.playerList.removeAll(this.isRestarting); // Paper
             try { Thread.sleep(100); } catch (InterruptedException ex) {} // CraftBukkit - SPIGOT-625 - give server at least a chance to send packets
         }
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index edf3f61593aef2ebc3bcf3785656d1009950104a..44eb6bfb79e156c47fddcbdbffe91c7144b39104 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1309,7 +1309,11 @@ public abstract class PlayerList {
     }
 
     public void saveAll(int interval) {
-        net.minecraft.server.MCUtil.ensureMain("Save Players" , () -> { // Paper - Ensure main
+        this.saveAll(interval, true); // MultiPaper
+    }
+
+    public void saveAll(int interval, boolean mainThread) {
+        Runnable savePlayers = () -> { // MultiPaper - allow off main
         MinecraftTimings.savePlayers.startTiming(); // Paper
         int numSaved = 0;
         long now = MinecraftServer.currentTick;
@@ -1322,7 +1326,13 @@ public abstract class PlayerList {
             // Paper end
         }
         MinecraftTimings.savePlayers.stopTiming(); // Paper
-        return null; }); // Paper - ensure main
+        };
+
+        if (mainThread) {
+            net.minecraft.server.MCUtil.ensureMain("Save Players" , savePlayers);
+        } else {
+            savePlayers.run(); // MultiPaper - allow off main
+        }
     }
 
     public UserWhiteList getWhiteList() {
