From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Wed, 26 Oct 2022 03:17:20 +0200
Subject: [PATCH] Supress spam messages.


diff --git a/src/main/java/net/minecraft/world/level/chunk/storage/EntityStorage.java b/src/main/java/net/minecraft/world/level/chunk/storage/EntityStorage.java
index ca81c3b396a303fda8eef8b8c1ea106ffecfc807..439b71faf18c4b13ed25e075a6b707effdbc956a 100644
--- a/src/main/java/net/minecraft/world/level/chunk/storage/EntityStorage.java
+++ b/src/main/java/net/minecraft/world/level/chunk/storage/EntityStorage.java
@@ -61,7 +61,7 @@ public class EntityStorage implements EntityPersistentStorage<Entity> {
                 })
                 .exceptionallyCompose(exception -> {
                     if (exception instanceof TimeoutException || exception.getCause() instanceof TimeoutException) {
-                        LOGGER.warn("Timed out reading " + world + "," + path + "," + cx + "," + cz + ", retrying...");
+//                        LOGGER.warn("Timed out reading " + world + "," + path + "," + cx + "," + cz + ", retrying...");
                         return (CompletionStage) loadAsync(pos); // Cast it away from what we want to get it to compile
                     }
                     return CompletableFuture.failedFuture(exception);
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index df163434fc5245d893d23ac375bc5f27d1315d4f..3ce1a53b6e21387878cb22c66aa5bf250c6b7ccd 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -360,7 +360,7 @@ public class MultiPaper {
         try {
             return readRegionFileAsync(world, path, cx, cz).get(20, TimeUnit.SECONDS);
         } catch (TimeoutException timeoutException) {
-            LOGGER.warn("Timed out reading " + world + "," + path + "," + cx + "," + cz + ", retrying...");
+//            LOGGER.warn("Timed out reading " + world + "," + path + "," + cx + "," + cz + ", retrying...");
             return readRegionFile(world, path, cx, cz);
         } catch (Exception e) {
             throw new RuntimeException(e);
diff --git a/src/main/java/puregero/multipaper/MultiPaperIO.java b/src/main/java/puregero/multipaper/MultiPaperIO.java
index b36b30f561f837105339109fa61552a8bdabd42b..58ba5b252b48ba176b09506086932af0b3438796 100644
--- a/src/main/java/puregero/multipaper/MultiPaperIO.java
+++ b/src/main/java/puregero/multipaper/MultiPaperIO.java
@@ -109,7 +109,7 @@ public class MultiPaperIO extends Thread implements IOHandler {
 
         future.thenApply(nbt -> false).completeOnTimeout(true, 15, TimeUnit.SECONDS).thenAccept(timedout -> {
             if (timedout) {
-                LOGGER.warn("Timed out reading " + key.getWorld() + "," + path + "," + key.getX() + "," + key.getZ() + ", retrying...");
+//                LOGGER.warn("Timed out reading " + key.getWorld() + "," + path + "," + key.getX() + "," + key.getZ() + ", retrying...");
                 readChunk(key, path).thenAccept(future::complete);
             }
         });
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/RequestChunkPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/RequestChunkPacket.java
index 9c4ff3f43b7bf419805129556a303417f4f9e036..52b05810b552b91a04417f2fd485e8ae9b544f66 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/RequestChunkPacket.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/RequestChunkPacket.java
@@ -137,7 +137,7 @@ public class RequestChunkPacket extends ExternalServerPacket {
         // Timeout instantly if this server is blocking our chunk loading, as this is probably also blocking their chunk loading
         .orTimeout(15, TimeUnit.SECONDS).exceptionally(throwable -> {
             if (throwable instanceof TimeoutException) {
-                LOGGER.warn("Timed out while sending chunk " + world + "," + cx + "," + cz);
+//                LOGGER.warn("Timed out while sending chunk " + world + "," + cx + "," + cz);
             } else {
                 LOGGER.warn("Error while sending chunk " + world + "," + cx + "," + cz, throwable);
             }
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/SendChunkPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/SendChunkPacket.java
index 8cbf1361cf46721c5bde0d9a067ea64dbdf47c2d..3dc39c7333568a765aad4a7db1e6c120d8c26f73 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/SendChunkPacket.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/SendChunkPacket.java
@@ -68,7 +68,7 @@ public class SendChunkPacket extends ExternalServerPacket {
             Consumer<DataInputStream> callback = connection.chunkCallbacks.remove(new ChunkKey(world, cx, cz));
             if (callback != null) {
                 if (data.length == 0) {
-                    LOGGER.warn(connection.externalServer.getName() + " sent us an empty chunk for " + world + "," + cx + "," + cz + ", force loading it from disk");
+//                    LOGGER.warn(connection.externalServer.getName() + " sent us an empty chunk for " + world + "," + cx + "," + cz + ", force loading it from disk");
                     MultiPaper.forceReadChunk(world, "region", cx, cz).thenAccept(data2 -> callback.accept(data2.length == 0 ? null : new DataInputStream(new InflaterInputStream(new ByteArrayInputStream(data2)))));
                 } else {
                     callback.accept(new DataInputStream(new ByteArrayInputStream(data)));
