From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Sun, 6 Nov 2022 17:01:24 +0100
Subject: [PATCH] Add option to disable stronghold generation.


diff --git a/src/main/java/net/minecraft/world/level/chunk/ChunkGenerator.java b/src/main/java/net/minecraft/world/level/chunk/ChunkGenerator.java
index 19ffd93b7bc745d9a6822f1e5642d2f640f61df7..8b378c9fada61995d65d250aaf2afb283a871de0 100644
--- a/src/main/java/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/src/main/java/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -230,6 +230,8 @@ public abstract class ChunkGenerator {
         return concentricringsstructureplacement.count() == 0 ? CompletableFuture.completedFuture(List.of()) : CompletableFuture.supplyAsync(Util.wrapThreadWithTaskName("placement calculation", () -> {
             Stopwatch stopwatch = Stopwatch.createStarted(Util.TICKER);
             List<ChunkPos> list = new ArrayList();
+            if (puregero.multipaper.config.MultiPaperConfiguration.get().optimizations.dontGenerateStrongholds) return list; // MultiPaper - optimize initial newchunk generation
+
             int i = concentricringsstructureplacement.distance();
             int j = concentricringsstructureplacement.count();
             int k = concentricringsstructureplacement.spread();
diff --git a/src/main/java/puregero/multipaper/config/MultiPaperConfiguration.java b/src/main/java/puregero/multipaper/config/MultiPaperConfiguration.java
index cf3acb949ca7f0bdc980c857b05c724ccc512e0c..b449cf0bf1bcfab7253d8d4487dc1b847316d679 100644
--- a/src/main/java/puregero/multipaper/config/MultiPaperConfiguration.java
+++ b/src/main/java/puregero/multipaper/config/MultiPaperConfiguration.java
@@ -64,6 +64,7 @@ public class MultiPaperConfiguration extends ConfigurationPart {
         public boolean shulkerBoxItemDropContentsWhenDestroyed = true;
         public boolean fastGetHealthSkippingBukkit = false;
         public boolean createRapidLogs = false;
+        public boolean dontGenerateStrongholds = false;
     }
 
     public SyncSettings syncSettings;
