From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Sat, 12 Nov 2022 12:32:00 +0100
Subject: [PATCH] "No chunk holder after ticket has been added" exception
 removed. xD


diff --git a/src/main/java/net/minecraft/server/level/ServerChunkCache.java b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
index e24d2987504733eded77d462faf305e4db8b61f8..10e004b69bcce5ccd9db6a1a9307dfb7e9d4ca7b 100644
--- a/src/main/java/net/minecraft/server/level/ServerChunkCache.java
+++ b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
@@ -582,7 +582,7 @@ public class ServerChunkCache extends ChunkSource {
                 gameprofilerfiller.pop();
                 if (this.chunkAbsent(playerchunk, l)) {
                     this.distanceManager.removeTicket(TicketType.REQUIRED_LOAD, chunkcoordintpair, l, identifier); // Paper
-                    throw (IllegalStateException) Util.pauseInIde(new IllegalStateException("No chunk holder after ticket has been added"));
+                    //throw (IllegalStateException) Util.pauseInIde(new IllegalStateException("No chunk holder after ticket has been added")); // MultiPaper - let's just ignore this instead of crashing...
                 }
             }
 
