From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Sat, 12 Nov 2022 15:41:33 +0100
Subject: [PATCH] Limit UseItemOnPackets.


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index e56c617fdbc15ab11693211a63aa5857fbef7c14..a0949e9275a6bdb147d4757c6881edc6f3c5fe5e 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1979,8 +1979,21 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
     }
     // Spigot end
 
+    long handleUseItemOn_start = 0;
+    long handleUseItemOn_totalTime = 0;
+
+    long HANDLE_USE_ITEM_ON_RESET = 1000000000; // 1s
+    long HANDLE_USE_ITEM_ON_LIMIT = 50000000; // 50ms
+
     @Override
     public void handleUseItemOn(ServerboundUseItemOnPacket packet) {
+        if (System.nanoTime() > HANDLE_USE_ITEM_ON_RESET + handleUseItemOn_start) {
+            handleUseItemOn_totalTime = 0;
+            handleUseItemOn_start = System.nanoTime();
+        }
+        long start = System.nanoTime();
+        if (handleUseItemOn_totalTime > 50000000) return;
+
         if (MultiPaperInventoryHandler.handleInteractEvent(this.player, packet)) return; // MultiPaper - send to external server if necessary
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.getLevel());
         if (this.player.isImmobile()) return; // CraftBukkit
@@ -2039,6 +2052,9 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
                 ServerGamePacketListenerImpl.LOGGER.warn("Rejecting UseItemOnPacket from {}: Location {} too far away from hit block {}.", new Object[]{this.player.getGameProfile().getName(), vec3d, blockposition});
             }
         }
+
+        System.out.println("TIME " + String.valueOf(handleUseItemOn_totalTime));
+        handleUseItemOn_totalTime += System.nanoTime() - start;
     }
 
     @Override
