From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Wed, 16 Nov 2022 02:24:20 +0100
Subject: [PATCH] Fix shutdown command.


diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index d6128e23935e1ec8773678a9116c5f0aec17a2d8..ba52ce56136f68190681e6cbfc5175e7a421a4c3 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -70,6 +70,7 @@ public class MultiPaper {
     public static boolean levelDatNeedsSaving = false;
     private static long last1Seconds = System.currentTimeMillis();
     private static long last10Seconds = System.currentTimeMillis();
+    public static boolean shutdownServer = false;
 
     public static MultiPaperConnection getConnection() {
         if (multiPaperConnection == null) {
@@ -188,6 +189,10 @@ public class MultiPaper {
         }
 
         MultiPaperPermissionSyncer.sync();
+
+        if (shutdownServer) {
+            MinecraftServer.getServer().halt(true);
+        }
     }
 
     public static void sendTickTime(long time, double tps) {
diff --git a/src/main/java/puregero/multipaper/MultiPaperConnection.java b/src/main/java/puregero/multipaper/MultiPaperConnection.java
index deee664d3270439565c4c689156d8f52c1f573c3..b5404f34ff4e314134ac7d733a9682df7aceb55c 100644
--- a/src/main/java/puregero/multipaper/MultiPaperConnection.java
+++ b/src/main/java/puregero/multipaper/MultiPaperConnection.java
@@ -454,7 +454,7 @@ public class MultiPaperConnection extends ServerBoundMessageHandler {
 
     @Override
     public void handle(ShutdownMessage message) {
-        Bukkit.shutdown();
+        MultiPaper.shutdownServer = true;
     }
 
     @Override
