From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Mon, 28 Nov 2022 01:59:18 +0100
Subject: [PATCH] Disable all profiling.


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 3a055c0b4e86864a6b6e65c2e20598a0295f7abf..90d4bf0fe1aed24aefb56b05b9b09ec5129d1669 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -2297,8 +2297,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     }
 
     public ProfilerFiller getProfiler() {
-        if (gg.airplane.AirplaneConfig.disableMethodProfiler) return net.minecraft.util.profiling.InactiveProfiler.INSTANCE;
-        return this.profiler;
+        return net.minecraft.util.profiling.InactiveProfiler.INSTANCE;
     }
 
     public abstract boolean isSingleplayerOwner(GameProfile profile);
diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index f245413e3eaa7356667420664775b2cd29aa7bb6..bc6ca38be626aee567638cf7532eee8c3bdaedfd 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -1498,8 +1498,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
     }
 
     public ProfilerFiller getProfiler() {
-        if (gg.airplane.AirplaneConfig.disableMethodProfiler) return net.minecraft.util.profiling.InactiveProfiler.INSTANCE; // Airplane
-        return (ProfilerFiller) this.profiler.get();
+        return net.minecraft.util.profiling.InactiveProfiler.INSTANCE;
     }
 
     public Supplier<ProfilerFiller> getProfilerSupplier() {
diff --git a/src/main/java/net/minecraft/world/level/PathNavigationRegion.java b/src/main/java/net/minecraft/world/level/PathNavigationRegion.java
index 138f5576a1df6714e754d9829f209f1a1678413f..e908a628314abb3a06be13f55b8601dce3e8d939 100644
--- a/src/main/java/net/minecraft/world/level/PathNavigationRegion.java
+++ b/src/main/java/net/minecraft/world/level/PathNavigationRegion.java
@@ -156,6 +156,6 @@ public class PathNavigationRegion implements BlockGetter, CollisionGetter {
     }
 
     public ProfilerFiller getProfiler() {
-        return this.level.getProfiler();
+        return net.minecraft.util.profiling.InactiveProfiler.INSTANCE;
     }
 }
