From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Wed, 30 Nov 2022 20:33:25 +0100
Subject: [PATCH] Limit TNT to 1000.


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 2067e943975eaf6ab6bdfc10a75ce9c0933112f2..4081b6ff19c83cc000ad2d7c76e788e63ef0a0f4 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -791,7 +791,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
             entityTimings.forEach((classType, totalTime) -> {
                 xymb.RapidLogger.i().logRaw("ENTITY;" + getWorld().getName() + ";" + classType.toString() + ";" + String.valueOf(entityCounts.get(classType)) + ";" + String.valueOf(totalTime));
 
-                boolean properType = AbstractHurtingProjectile.class.isAssignableFrom(classType) || ThrowableProjectile.class.isAssignableFrom(classType);
+                boolean properType = AbstractHurtingProjectile.class.isAssignableFrom(classType) || ThrowableProjectile.class.isAssignableFrom(classType) || net.minecraft.world.entity.item.PrimedTnt.class.isAssignableFrom(classType);
 
                 if (properType && totalTime > TOTAL_TIME_PER_ENTITY_TYPE && entityCounts.get(classType).intValue() > SPAM_LIMIT_PER_ENTITY_TYPE) {
                     long start = System.nanoTime();
