From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Mon, 28 Nov 2022 08:05:27 +0100
Subject: [PATCH] Skip unnecessary NBT copy when transferring items through
 hoppers.


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 4b6e8b234120ea66a09a779dc7f1c98465ce5161..9baea44c3ca98a695ef1317147188cdcbb39978b 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -708,6 +708,19 @@ public final class ItemStack {
     }
 
     // MultiPaper start
+    public ItemStack rapidCloneItemStack(boolean origItem) {
+        if (!origItem && this.isEmpty()) {
+            return ItemStack.EMPTY;
+        } else {
+            ItemStack itemstack = new ItemStack(origItem ? this.item : this.getItem(), this.count);
+
+            itemstack.setPopTime(this.getPopTime());
+            itemstack.tag = this.tag;
+
+            return itemstack;
+        }
+    }
+
     public static boolean rapidItemStackMatch(ItemStack left, ItemStack right) {
         if(!left.isEmpty() && !right.isEmpty()) {
             if(left.getItem() instanceof BlockItem && right.getItem() instanceof BlockItem) {
diff --git a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index 8cd86e4bed7c9d890ed4c4f002cbc574e70b53f8..04466d7d91ffdcf7ce5a3c7d21b3af06173d7a6e 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -250,7 +250,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                 final ItemStack itemstack2 = addItem(hopper, destination, itemstack, enumdirection);
                 final int remaining = itemstack2.getCount();
                 if (remaining != moved) {
-                    origItemStack = origItemStack.cloneItemStack(true);
+                    origItemStack = origItemStack.rapidCloneItemStack(true); // MultiPaper
                     origItemStack.setCount(origCount);
                     if (!origItemStack.isEmpty()) {
                         origItemStack.setCount(origCount - moved + remaining);
@@ -302,7 +302,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         final ItemStack itemstack2 = addItem(iinventory, ihopper, itemstack, null);
         final int remaining = itemstack2.getCount();
         if (remaining != moved) {
-            origItemStack = origItemStack.cloneItemStack(true);
+            origItemStack = origItemStack.rapidCloneItemStack(true); // MultiPaper
             origItemStack.setCount(origCount);
             if (!origItemStack.isEmpty()) {
                 origItemStack.setCount(origCount - moved + remaining);
