From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Wed, 30 Nov 2022 21:02:33 +0100
Subject: [PATCH] Packet limiter for handleContainerClick.


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 4ce04de910698f993b2e0907869273de448725e6..d45ed480f870550df020e36c27527ad1ccc5c2ce 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1983,17 +1983,17 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
     long handleUseItemOn_start = 0;
     long handleUseItemOn_totalTime = 0;
 
-    long HANDLE_USE_ITEM_ON_RESET = 1000000000; // 1s
-    long HANDLE_USE_ITEM_ON_LIMIT = 50000000; // 50ms
-
     @Override
     public void handleUseItemOn(ServerboundUseItemOnPacket packet) {
-        if (System.nanoTime() > HANDLE_USE_ITEM_ON_RESET + handleUseItemOn_start) {
+        if (System.nanoTime() > HANDLE_RESET + handleUseItemOn_start) {
             handleUseItemOn_totalTime = 0;
             handleUseItemOn_start = System.nanoTime();
         }
+        if (handleUseItemOn_totalTime > HANDLE_LIMIT) {
+            xymb.RapidLogger.i().logRaw("LAG_HANDLE_USE_ITEM_ON_PACKET;" + this.player.getName());
+            return;
+        }
         long start = System.nanoTime();
-        if (handleUseItemOn_totalTime > 50000000) return;
 
         if (MultiPaperInventoryHandler.handleInteractEvent(this.player, packet)) return; // MultiPaper - send to external server if necessary
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.getLevel());
@@ -3147,8 +3147,24 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
         this.player.doCloseContainer();
     }
 
+    long handleContainerClick_start = 0;
+    long handleContainerClick_totalTime = 0;
+
+    long HANDLE_RESET = 1000000000; // 1s
+    long HANDLE_LIMIT = 50000000; // 50ms
+
     @Override
     public void handleContainerClick(ServerboundContainerClickPacket packet) {
+        if (System.nanoTime() > HANDLE_RESET + handleUseItemOn_start) {
+            handleContainerClick_totalTime = 0;
+            handleContainerClick_start = System.nanoTime();
+        }
+        if (handleContainerClick_totalTime > HANDLE_LIMIT) {
+            xymb.RapidLogger.i().logRaw("LAG_HANDLE_CONTAINER_CLICK_PACKET;" + this.player.getName());
+            return;
+        }
+        long start = System.nanoTime();
+
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.getLevel());
         if (MultiPaperInventoryHandler.handleContainerEvent(this.player, packet)) return; // MultiPaper
         if (this.player.isImmobile()) return; // CraftBukkit
