From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Sun, 4 Dec 2022 03:06:35 +0100
Subject: [PATCH] Remove memory leak prevention.


diff --git a/src/main/java/net/minecraft/server/level/DistanceManager.java b/src/main/java/net/minecraft/server/level/DistanceManager.java
index ede901293c99677af7e8c22bff24ff83a7d7284a..b77b553ca7e96ded7ac310bb07e3f9edbddb916a 100644
--- a/src/main/java/net/minecraft/server/level/DistanceManager.java
+++ b/src/main/java/net/minecraft/server/level/DistanceManager.java
@@ -186,7 +186,7 @@ public abstract class DistanceManager {
         //this.f.a(); // Paper - no longer used
         //this.tickingTicketsTracker.runAllUpdates(); // Paper - no longer used
         org.spigotmc.AsyncCatcher.catchOp("DistanceManagerTick"); // Paper
-        if (chunkStorage.queueSorter.totalSize() > 5000) return false; // MultiPaper - prevent memory leak
+//        if (chunkStorage.queueSorter.totalSize() > 5000) return false; // MultiPaper - prevent memory leak
         // this.playerTicketManager.runAllUpdates(); // Paper - no longer used
         boolean flag = this.ticketLevelPropagator.propagateUpdates(); // Paper - replace ticket level propagator
 
