From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Fri, 2 Dec 2022 04:55:01 +0100
Subject: [PATCH] Fix SendChunkPacket crash.


diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/SendChunkPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/SendChunkPacket.java
index 8cbf1361cf46721c5bde0d9a067ea64dbdf47c2d..9891f1557d9b3d3f64912e095d54b1042d47a4fb 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/SendChunkPacket.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/SendChunkPacket.java
@@ -69,7 +69,7 @@ public class SendChunkPacket extends ExternalServerPacket {
             if (callback != null) {
                 if (data.length == 0) {
                     LOGGER.warn(connection.externalServer.getName() + " sent us an empty chunk for " + world + "," + cx + "," + cz + ", force loading it from disk");
-                    MultiPaper.forceReadChunk(world, "region", cx, cz).thenAccept(data2 -> callback.accept(data2.length == 0 ? null : new DataInputStream(new InflaterInputStream(new ByteArrayInputStream(data2)))));
+                    MultiPaper.forceReadChunk(world, "region", cx, cz).thenAccept(data2 -> callback.accept(data2.length == 0 ? null : new DataInputStream(new ByteArrayInputStream(data2))));
                 } else {
                     callback.accept(new DataInputStream(new ByteArrayInputStream(data)));
                 }
