From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Sat, 3 Dec 2022 09:21:35 +0100
Subject: [PATCH] Fix "player" is null spam message.


diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/AdvancementGrantProgressPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/AdvancementGrantProgressPacket.java
index e1130055ba1cb66678d8e6bbeecef5dc98aea7f5..e35f9fa1e2e86f64bb9025343f887c208a5eb6d7 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/AdvancementGrantProgressPacket.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/AdvancementGrantProgressPacket.java
@@ -42,7 +42,7 @@ public class AdvancementGrantProgressPacket extends ExternalServerPacket {
         MultiPaper.runSync(() -> {
             updatingAdvancements = true;
             ServerPlayer player = MinecraftServer.getServer().getPlayerList().getPlayer(uuid);
-            player.getAdvancements().award(player.getServer().getAdvancements().getAdvancement(new ResourceLocation(advancement)), criterion);
+            if (player != null) player.getAdvancements().award(player.getServer().getAdvancements().getAdvancement(new ResourceLocation(advancement)), criterion);
             updatingAdvancements = false;
         });
     }
