From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Sun, 4 Dec 2022 17:14:23 +0100
Subject: [PATCH] Removed a redundant copy from FriendlyByteBuf


diff --git a/src/main/java/net/minecraft/network/FriendlyByteBuf.java b/src/main/java/net/minecraft/network/FriendlyByteBuf.java
index 970ed3671b7c6bf901a4ee7bf39887d976a5edce..46d89d3dc9d607ca171db3d819b6c4d6d420efc9 100644
--- a/src/main/java/net/minecraft/network/FriendlyByteBuf.java
+++ b/src/main/java/net/minecraft/network/FriendlyByteBuf.java
@@ -580,7 +580,7 @@ public class FriendlyByteBuf extends ByteBuf {
 
             if (item.canBeDepleted() || item.shouldOverrideMultiplayerNbt()) {
                 // Spigot start - filter
-                stack = stack.copy();
+                // stack = stack.copy(); // MultiPaper - seems unnecessary
                 // CraftItemStack.setItemMeta(stack, CraftItemStack.getItemMeta(stack)); // Paper - This is no longer with raw NBT being handled in metadata
                 // Spigot end
                 nbttagcompound = stack.getTag();
