From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Wed, 14 Dec 2022 03:30:34 +0100
Subject: [PATCH] Some random changes.


diff --git a/src/main/java/net/minecraft/world/ContainerHelper.java b/src/main/java/net/minecraft/world/ContainerHelper.java
index 050b82225a20d843661ee963f7ab9d92d1277a1d..7c5e37a22d79eccce160cda1030fdd47e7ea19ce 100644
--- a/src/main/java/net/minecraft/world/ContainerHelper.java
+++ b/src/main/java/net/minecraft/world/ContainerHelper.java
@@ -67,6 +67,7 @@ public class ContainerHelper {
                     stacks.set(j, ItemStack.ofRapidUnsafe(compoundTag));
                 }
             }
+            nbt.remove("NewItems");
         }
     }
 
diff --git a/src/main/java/puregero/multipaper/MultiPaperChunkHandler.java b/src/main/java/puregero/multipaper/MultiPaperChunkHandler.java
index 99b7264686793ee5af8797c8c3d7dc0e6e6f0e8d..09f86aa92d697c8d8359d907211754829634c23a 100644
--- a/src/main/java/puregero/multipaper/MultiPaperChunkHandler.java
+++ b/src/main/java/puregero/multipaper/MultiPaperChunkHandler.java
@@ -120,7 +120,9 @@ public class MultiPaperChunkHandler {
         if (blockUpdateChunk == null) { // Don't broadcast the update to other servers if we're handling an update
             for (ExternalServer externalServer : chunk.externalSubscribers) {
                 if (!externalServer.isMe()) {
+                    try {
                     externalServer.getConnection().send(new SendUpdatePacket(chunkHolder.getWorld().uuid, packet));
+                    } catch (java.lang.NullPointerException ex) {}
                 }
             }
         }
diff --git a/src/main/java/xymb/DeduplicatedNBTTag.java b/src/main/java/xymb/DeduplicatedNBTTag.java
index 097a4d7d9f128e2e59fd3d93620ebd77bb9a99f0..11f01c1e4e9c76065cfdbb9f1137e544f1eb1ba7 100644
--- a/src/main/java/xymb/DeduplicatedNBTTag.java
+++ b/src/main/java/xymb/DeduplicatedNBTTag.java
@@ -20,8 +20,7 @@ public class DeduplicatedNBTTag implements Tag {
     }
 
     public void write(DataOutput output) throws IOException {
-//        tag.write(output);
-//        throw new IOException("DeduplicatedNBTTag is not supposed to be written!");
+        throw new IOException("DeduplicatedNBTTag is not supposed to be written!");
     }
 
     public String toString() {
