From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Thu, 15 Dec 2022 17:29:39 +0100
Subject: [PATCH] Increase "Timed out while sending chunk" to 30s.


diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/RequestChunkPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/RequestChunkPacket.java
index 9c4ff3f43b7bf419805129556a303417f4f9e036..90939c7c61d70bb14269e1356b502fdd9e6a51ee 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/RequestChunkPacket.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/RequestChunkPacket.java
@@ -135,7 +135,7 @@ public class RequestChunkPacket extends ExternalServerPacket {
             }
         })
         // Timeout instantly if this server is blocking our chunk loading, as this is probably also blocking their chunk loading
-        .orTimeout(15, TimeUnit.SECONDS).exceptionally(throwable -> {
+        .orTimeout(30, TimeUnit.SECONDS).exceptionally(throwable -> {
             if (throwable instanceof TimeoutException) {
                 LOGGER.warn("Timed out while sending chunk " + world + "," + cx + "," + cz);
             } else {
