From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Mon, 24 Oct 2022 02:54:26 +0200
Subject: [PATCH] Disable blocking chunk load on entity move.


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 71e2bbe10fa4bdb3410bf77e06eb4c899bf235b1..10bd701f7af1b91409cd4796130afacf3331d598 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1989,6 +1989,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         this.yo = y;
         this.zo = d4;
         this.setPos(d3, y, d4);
+        if (!((ServerChunkCache) this.level.getChunkSource()).isChunkLoaded((int) Math.floor(this.getX()) >> 4, (int) Math.floor(this.getZ()) >> 4)) return; // MultiPaper
         if (this.valid) this.level.getChunk((int) Math.floor(this.getX()) >> 4, (int) Math.floor(this.getZ()) >> 4); // CraftBukkit
     }
 
