From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivan Pekov <ivan@mrivanplays.com>
Date: Mon, 23 Jan 2023 19:05:23 +0200
Subject: [PATCH] Improve Entity#stopRiding


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index c7ea15ca0c6594e832d58ec4aed5c86d0d2052dd..965312b4321a12da54d85b3761b696b2c806b287 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2837,7 +2837,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     public void stopRiding(boolean suppressCancellation) {
         // Paper end
         if (this.vehicle != null) {
-            boolean wasControllingPassenger = MultiPaperEntitiesHandler.getControllingPassenger(this.vehicle) == this; // MultiPaper
+            var controllingPassenger = MultiPaperEntitiesHandler.getControllingPassenger(this.vehicle);
+            boolean wasControllingPassenger = controllingPassenger == this; // MultiPaper
+            if (!wasControllingPassenger && controllingPassenger.getUUID().equals(this.getUUID())) {
+                wasControllingPassenger = true;
+            }
             Entity entity = this.vehicle;
 
             this.vehicle = null;
@@ -2845,9 +2849,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
             // MultiPaper start - inform other servers that the player is no longer riding the entity and hand them the entity's data
             LevelChunk chunk = level.getChunkIfLoaded(entity.blockPosition);
-            if (vehicle == null && MultiPaper.isRealPlayer(this) && wasControllingPassenger && !EntityRemovePacket.removedEntities.containsKey(entity.getRootVehicle().getUUID())) {
+            if (chunk != null && vehicle == null && MultiPaper.isRealPlayer(this) && wasControllingPassenger && !EntityRemovePacket.removedEntities.containsKey(entity.getRootVehicle().getUUID())) {
                 // Any server might now be controlling this vehicle
-                Set<ExternalServer> servers = new LinkedHashSet<>(chunk == null ? Collections.emptyList() : chunk.externalEntitiesSubscribers);
+                Set<ExternalServer> servers = new LinkedHashSet<>(chunk.externalEntitiesSubscribers);
 
                 if (MultiPaperEntitiesHandler.getControllingPassenger(this.vehicle) instanceof ExternalPlayer newController) {
                     // Make sure we send the vehicle to the new controlling passenger, even if they don't have the chunk loaded
