From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivan Pekov <ivan@mrivanplays.com>
Date: Mon, 23 Jan 2023 20:06:56 +0200
Subject: [PATCH] Do not store removed entities

idk how impactful this is

diff --git a/src/main/java/net/minecraft/world/level/chunk/storage/EntityStorage.java b/src/main/java/net/minecraft/world/level/chunk/storage/EntityStorage.java
index 707374502a35ced453bc861a0a2cdba7ac766d4f..61555915c1755eb8691638e3462a897b782a4191 100644
--- a/src/main/java/net/minecraft/world/level/chunk/storage/EntityStorage.java
+++ b/src/main/java/net/minecraft/world/level/chunk/storage/EntityStorage.java
@@ -25,6 +25,7 @@ import net.minecraft.world.level.entity.EntityPersistentStorage;
 import org.slf4j.Logger;
 import puregero.multipaper.MultiPaper;
 import puregero.multipaper.MultiPaperEntitiesHandler;
+import puregero.multipaper.externalserverprotocol.EntityRemovePacket;
 
 public class EntityStorage implements EntityPersistentStorage<Entity> {
     private static final Logger LOGGER = LogUtils.getLogger();
@@ -139,6 +140,10 @@ public class EntityStorage implements EntityPersistentStorage<Entity> {
             ListTag listTag = new ListTag();
             final java.util.Map<net.minecraft.world.entity.EntityType<?>, Integer> savedEntityCounts = new java.util.HashMap<>(); // Paper
             dataList.getEntities().forEach((entity) -> {
+                // do not store removed
+                if (EntityRemovePacket.removedEntities.containsKey(entity.getUUID())) {
+                    return;
+                }
                 // Paper start
                 final EntityType<?> entityType = entity.getType();
                 final int saveLimit = this.level.paperConfig().chunks.entityPerChunkSaveLimit.getOrDefault(entityType, -1);
