From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Thu, 12 Jan 2023 23:49:26 +0100
Subject: [PATCH] Log laggy JUMBO_PACKETS in the console.


diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/SendChunkPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/SendChunkPacket.java
index bed47074d3d387d5442cf5b953af644ae0c6076d..7b6f86f4a314c4628071bdf54b5aa014b26c8d34 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/SendChunkPacket.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/SendChunkPacket.java
@@ -41,7 +41,11 @@ public class SendChunkPacket extends ExternalServerPacket {
             if (tag == null) {
                 data = new byte[0];
             } else {
+                long startTime = System.nanoTime();
                 data = MultiPaper.nbtToBytes(tag);
+                if (System.nanoTime() - startTime > 10000000) {
+                    System.out.println("SEND_CHUNK_JUMBO_PACKET " + ((System.nanoTime() - startTime) / 1000000.) + " " + world + " " + cx + " " + cz);
+                }
             }
         } catch (IOException e) {
             throw new RuntimeException(e);
